app/controllers/passwordless/sessions_controller.rb

```diff
@@ -42,28 +42,51 @@ def create
     def show
       # Make it "slow" on purpose to make brute-force attacks more of a hassle
       BCrypt::Password.create(params[:token])
-
-      destination =
-        Passwordless.redirect_back_after_sign_in &&
-        reset_passwordless_redirect_location!(authenticatable_class)
-
       sign_in passwordless_session
 
-      redirect_to destination || main_app.root_path
+      redirect_to passwordless_success_redirect_path
     rescue Errors::TokenAlreadyClaimedError
       flash[:error] = I18n.t(".passwordless.sessions.create.token_claimed")
-      redirect_to main_app.root_path
+      redirect_to passwordless_failure_redirect_path
     rescue Errors::SessionTimedOutError
       flash[:error] = I18n.t(".passwordless.sessions.create.session_expired")
-      redirect_to main_app.root_path
+      redirect_to passwordless_failure_redirect_path
     end
 
     # match '/sign_out', via: %i[get delete].
     #   Signs user out. Redirects to root_path
     # @see ControllerHelpers#sign_out
     def destroy
       sign_out authenticatable_class
-      redirect_to main_app.root_path
+      redirect_to passwordless_sign_out_redirect_path
+    end
+
+    protected
+
+    def passwordless_sign_out_redirect_path
+      Passwordless.sign_out_redirect_path
+    end
+
+    def passwordless_failure_redirect_path
+      Passwordless.failure_redirect_path
+    end
+
+    def passwordless_success_redirect_path
+      return Passwordless.success_redirect_path unless Passwordless.redirect_back_after_sign_in
+
+      begin
+        uri = URI.parse(params[:destination_path])
+        query_redirect_path = StringIO.new
+        query_redirect_path << uri.path
+        query_redirect_path << '?' << uri.query if uri.query.present?
+        query_redirect_path << '#' << uri.fragment if uri.fragment.present?
+        query_redirect_path = query_redirect_path.string
```

Comment:
I think this could possibly be simplified to:

```ruby
  def query_redirect_path
    query_redirect_uri = URI(params[:destination_path])
    query_redirect_uri.to_s if query_redirect_uri.host == URI(request.url).host
  rescue URI::InvalidURIError
    nil
  end
```

Inspired by https://github.com/rails/rails/blob/66cabeda2c46c582d19738e1318be8d59584cc5b/actionpack/lib/action_controller/metal/redirecting.rb#L127-L131

What do you think?

Reply:
Absolutely, I've pushed a change for this and updated tests.